File: src/main/java/com/dude/dms/ui/components/dialogs/ChangelogDialog.java
package com.dude.dms.ui.components.dialogs;

import com.dude.dms.backend.data.updater.Changelog;
import com.dude.dms.backend.service.ChangelogService;
import com.dude.dms.updater.UpdateChecker;
import com.vaadin.flow.component.button.Button;
import com.vaadin.flow.component.button.ButtonVariant;
import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.component.textfield.TextArea;

import java.util.List;
import java.util.stream.Collectors;

public class ChangelogDialog extends Dialog {

    public ChangelogDialog(ChangelogService changelogService, UpdateChecker updateChecker) {
        setWidth("70vw");
        setHeight("70vh");
        Button b = new Button("Check for updates", e -> updateChecker.check(false));
        b.addThemeVariants(ButtonVariant.LUMO_PRIMARY);
        add(b);
        List<Changelog> changelogs = changelogService.findAll().stream().sorted((o1, o2) -> o2.getPublished().compareTo(o1.getPublished())).collect(Collectors.toList());
        for (Changelog changelog : changelogs) {
            TextArea area = new TextArea(changelog.getVersion(), changelog.getBody(), "");
            area.setReadOnly(true);
            area.setWidthFull();
            add(area);
        }
    }
}

File: src/main/java/com/dude/dms/ui/components/dialogs/ChangelogDialog.kt
package com.dude.dms.ui.components.dialogs

import com.dude.dms.brain.t
import com.dude.dms.extensions.changelogService
import com.dude.dms.extensions.updateChecker
import com.github.mvysny.karibudsl.v10.button
import com.github.mvysny.karibudsl.v10.onLeftClick
import com.github.mvysny.karibudsl.v10.textArea
import com.vaadin.flow.component.button.ButtonVariant
import com.vaadin.flow.component.icon.VaadinIcon

class ChangelogDialog : DmsDialog(t("changelog"), 70, 70) {

    init {
        button(t("check"), VaadinIcon.REFRESH.create()) {
            onLeftClick { updateChecker.check(false) }
            addThemeVariants(ButtonVariant.LUMO_PRIMARY)
        }

        changelogService.findAll().sortedBy { it.published }.reversed().forEach { changelog ->
            textArea(changelog.version) {
                setWidthFull()
                value = changelog.body
                isReadOnly = true
            }
        }
    }
}